src/clj/puppetlabs/services/ca/certificate_authority_service.clj

```diff
@@ -45,20 +40,29 @@
          auth-handler
          puppet-version)
        {:normalize-request-uri true})
-     (when (not= ca-crl-file host-crl-file)
-       (watch-protocol/add-watch-dir! watcher
-                                      (fs/parent ca-crl-file)
-                                      {:recursive true})
-       (watch-protocol/add-callback!
-        watcher
-        (fn [events]
-          (when (some #(and (:changed-path %)
-                            (= (.getCanonicalPath (:changed-path %))
-                               ca-crl-file))
-                      events)
-            (ca/retrieve-ca-crl! ca-crl-file host-crl-file)))))
-     (assoc context :auth-handler auth-handler
-                    :watcher watcher)))
+       (if-let [filesystem-watch-service
+                 (tk-services/maybe-get-service this :FilesystemWatchService)]
+         (let [ca-crl-file (.getCanonicalPath (fs/file
+                                           (get-in-config
+                                             [:puppetserver :cacrl])))
+               host-crl-file (.getCanonicalPath (fs/file
+                                             (get-in-config
+                                               [:puppetserver :hostcrl])))
+               watcher (watch-protocol/create-watcher filesystem-watch-service)]
+           (when (not= ca-crl-file host-crl-file)
+             (watch-protocol/add-watch-dir! watcher
+                                             (fs/parent ca-crl-file)
+                                             {:recursive true})
+             (watch-protocol/add-callback!
+               watcher
+               (fn [events]
+                 (when (some #(and (:changed-path %)
+                                   (= (.getCanonicalPath (:changed-path %))
+                                     ca-crl-file))
+                             events)
+                   (ca/retrieve-ca-crl! ca-crl-file host-crl-file)))))
+           (assoc context' :watcher watcher))
```

Comment:
maybe because we have to return the context no matter what and since we end up with 3 different possible conditional forks otherwise we end up repeating ourselves a lot if we don't